python/cudf/cudf/core/_internals/where.py

```diff
@@ -0,0 +1,383 @@
+# Copyright (c) 2021, NVIDIA CORPORATION.
+
+import warnings
+from typing import Any, Optional, Tuple, Union, cast
+
+import numpy as np
+import pandas as pd
+
+import cudf
+from cudf._typing import ColumnLike, ScalarLike
+from cudf.core.column import ColumnBase
+from cudf.core.dataframe import DataFrame
+from cudf.core.frame import Frame
+from cudf.core.index import Index
+from cudf.core.series import Series
+
+
+def _normalize_scalars(col: ColumnBase, other: ScalarLike) -> ScalarLike:
+    """
+    Try to normalize scalar values as per col dtype
+    """
+    if (isinstance(other, float) and not np.isnan(other)) and (
+        col.dtype.type(other) != other
+    ):
+        raise TypeError(
+            f"Cannot safely cast non-equivalent "
+            f"{type(other).__name__} to {col.dtype.name}"
+        )
+
+    return cudf.Scalar(other, dtype=col.dtype if other is None else None)
+
+
+def _check_and_cast_columns_with_other(
+    source_col: ColumnBase,
+    other: Union[ScalarLike, ColumnBase],
+    inplace: bool,
+) -> Tuple[ColumnBase, Union[ScalarLike, ColumnBase]]:
+    """
+    Returns type-casted column `source_col` & scalar `other_scalar`
+    based on `inplace` parameter.
+    """
+    if cudf.utils.dtypes.is_categorical_dtype(source_col.dtype):
+        return source_col, other
+
+    if cudf.utils.dtypes.is_scalar(other):
+        device_obj = _normalize_scalars(source_col, other)
+    else:
+        device_obj = other
+
+    if other is None:
+        return source_col, device_obj
+    elif cudf.utils.dtypes.is_mixed_with_object_dtype(device_obj, source_col):
+        raise TypeError(
+            "cudf does not support mixed types, please type-cast "
+            "the column of dataframe/series and other "
+            "to same dtypes."
+        )
+    if inplace:
+        if not cudf.utils.dtypes._can_cast(device_obj.dtype, source_col.dtype):
+            warnings.warn(
+                f"Type-casting from {device_obj.dtype} "
+                f"to {source_col.dtype}, there could be potential data loss"
+            )
+        return source_col, device_obj.astype(source_col.dtype)
+    else:
+        if (
+            cudf.utils.dtypes.is_scalar(other)
+            and pd.api.types.is_numeric_dtype(source_col.dtype)
```

Comment:
Can we use cudf's `is_numeric_dtype` here instead?

Reply:
Is it `cudf.util.dtypes.is_numerical_dtype`?

Reply:
Yup, sorry - that's the one I meant

Reply:
I think that API is still not equipped to handle `dtype` objects.

Reply:
I introduced some improvements recently where I think now it should:

https://github.com/rapidsai/cudf/blob/branch-0.19/python/cudf/cudf/utils/dtypes.py#L146-L153

Reply:
Cool, thanks. Changed it 👍